pulp_docker/app/models.py

```diff
@@ -187,12 +202,83 @@ class DockerPublisher(Publisher):
 class DockerRemote(Remote):
     """
     A Remote for DockerContent.
-
-    Define any additional fields for your new importer if needed.
     """
 
+    upstream_name = models.CharField(max_length=255, db_index=True)
+
     TYPE = 'docker'
 
+    def __init__(self, *args, **kwargs):
+        """
+        Create a remote instance that includes a token and token_lock.
+        """
+        super().__init__(*args, **kwargs)
+        self.token = {'token': None}
+        self._token_lock = None
```

Comment:
It's my understanding that - In _asyncio_, context switching only happens at _await_ statements.  What mutual exclusion is this lock intended to provide?

Reply:
I was concerned about the possibility of many instances of the downloader being used at once. For a given remote, only 1 token should be retrieved at a time. Also, when I was writing this, I had hoped that the TokenAuthDownloader could be used for other token auth situations, not just Docker, but I have added other features to it that are (barely) docker specific. I agree, I'd rather have this on the TokenAuthDownloader, which I can test now that the rest of sync is working.

Reply:
Agreed, the lock is needed.  Suggest that is would be more appropriate as a class attribute on the downloader.
```
class TokenAuthHttpDownloader(HttpDownloader):
    _lock = asyncio.RLock()

```

Reply:
Asyncio does not have an RLock. I think this is still going to work.